All KIC Reference pages (#1121)
Kong/developer.konghq.com: app/kubernetes-ingress-controller/reference/feature-gates.md

---

> title: Feature Gates

Reviewer: ```suggestion
title: Feature gates
```

description: |
  Learn how to customize {{ site.kic_product_name }}'s behavior using feature flags

breadcrumbs:
  - /kubernetes-ingress-controller/
  - index: kubernetes-ingress-controller
    section: Reference

content_type: reference
layout: reference

products:
  - kic

works_on:
  - on-prem
  - konnect

> ---

Reviewer: ```suggestion
related_resources:
  - text: Stages of software availability
    url: /stages-of-software-availability/
  - text: Kubernetes
    url: https://kubernetes.io
  - text: Kubernetes feature gates
    url: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
  - text: Feature stages
    url: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages
---
```
The links from the intro + link to availability definitions

> Feature gates enables contributors to add and manag`e new (and potentially experimental) functionality to the KIC in a controlled manner. The features will be "hidden" until generally available (GA) and the progress and maturity of features on their path to GA will be documented. Feature gates also create a clear path for deprecating features.

Reviewer: ```suggestion
Feature gates enable contributors to add and manage new (and potentially experimental) functionality to the {{site.kic_product_name}} in a controlled way. The features will be "hidden" until they are generally available (GA) and the progress and maturity of features on their path to GA will be documented. Feature gates also create a clear path for deprecating features.
```

Upstream [Kubernetes](https://kubernetes.io) includes [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/), which enable or disable features with flags and track the maturity of a feature using [feature stages](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages).
The same definitions of `feature gates` and `feature stages` from upstream Kubernetes are used to define KIC's list of features.

> * [Kubernetes](https://kubernetes.io)
> * [Kubernetes feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
> * [Feature stages](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages)

Reviewer: ```suggestion
```
These should really be in the related_resources, especially since they're just repeats of links in the previous paragraph.

    stage: GA
    since: 3.4.0
    until: TBD
  - feature: CombinedServicesFromDifferentHTTPRoutes
    default: '`false`'
    stage: Alpha
    since: 3.4.0
    until: TBD
{% endtable %}

## Using feature gates

To enable feature gates, provide the `--feature-gates` flag when launching KIC, or set the `CONTROLLER_FEATURE_GATES` environment variable.

Feature gates consist of a comma-delimited set of `key=value` pairs. For example, if you wanted to enable `FillIDs` and `RewriteURIs`, you'd set `CONTROLLER_FEATURE_GATES=FillIDs=true,RewriteURIs=true`.

To enable features via Helm, set the following in your `values.yaml`:

{% navtabs chart %}
{% navtab "kong/ingress" %}
```yaml
controller:
  ingressController:
    env:
      feature_gates: FillIDs=true,RewriteURIs=true
```
{% endnavtab %}
{% navtab "kong/kong" %}
```yaml
ingressController:
  env:
    feature_gates: FillIDs=true,RewriteURIs=true
```
{% endnavtab %}
{% endnavtabs %}

To test a feature gate in an existing deployment, use `kubectl set env`.

```bash
kubectl set env -n kong deployment/kong-controller \
  CONTROLLER_FEATURE_GATES="FillIDs=true,RewriteURIs=true" \
  -c ingress-controller
```

## Feature gate availability

> * The **since** and **until** rows in the table above refer to [KIC Releases](https://github.com/Kong/kubernetes-ingress-controller/releases).

Reviewer: ```suggestion
* The **since** and **until** rows in the [feature gates table](#available-feature-gates) refer to [KIC Releases](https://github.com/Kong/kubernetes-ingress-controller/releases).
```

* Features that are currently in alpha or beta states may become deprecated at any time. Deprecated features are removed during the next minor release.

> * Until a feature becomes GA, there are no guarantees that will continue being available. For more information, see the [changelog](https://github.com/Kong/kubernetes-ingress-controller/blob/main/CHANGELOG.md).

Reviewer: ```suggestion
* Until a feature becomes GA, there are no guarantees that it will continue being available. For more information, see the [changelog](https://github.com/Kong/kubernetes-ingress-controller/blob/main/CHANGELOG.md).
```

{:.warning}
>**Important:** To avoid disrupting your services, consider not using features until they have reached GA status.

> ## Feature gate details

Reviewer: Why does this only contain two of the above feature gates?

Author: Resolving for now to keep things moving. We should probably cover all feature gates.

Captured as https://github.com/Kong/developer.konghq.com/issues/1187

### SanitizeKonnectConfigDumps

> The `SanitizeKonnectConfigDumps` feature enables the sanitization of configuration dumps that are sent to Konnect.

Reviewer: ```suggestion
The `SanitizeKonnectConfigDumps` feature enables the sanitization of configuration dumps that are sent to {{site.konnect_short_name}}.
```

This means {{site.kic_product_name}} will obfuscate all sensitive information that your Kong config contains, such as
private keys in `Certificate` entities and `Consumer` entities' credentials.

{:.warning}
> **Warning:** `KongPlugin`'s and `KongClusterPlugin`'s `config` fields are not sanitized. If you have sensitive information
> in your `KongPlugin`'s `config` field, it will be sent to Konnect as is. To avoid this, use the
> [KongVault](/kubernetes-ingress-controller/{{page.release}}/reference/custom-resources/#kongvault) resource.

### CombinedServicesFromDifferentHTTPRoutes

> The `CombinedServicesFromDifferentHTTPRoutes` feature enables translating `HTTPRoute` rules
> with the same set of backends (combination of namespace, name, port and weight) from different `HTTPRoute`s in the same namespace
> into a single {{site.base_gateway}} service. Enabling the feature gate can reduce the number of translated {{site.base_gateway}} services.
> 
> The names of {{site.base_gateway}} services will change if the feature gate is enabled.
> You can refer to the [reference page](/kubernetes-ingress-controller/{{page.release}}/reference/combined-services-from-different-httproutes) for further details.

Reviewer: ```suggestion
The `CombinedServicesFromDifferentHTTPRoutes` feature enables translating `HTTPRoute` rules
with the same set of backends (combination of namespace, name, port, and weight) from different `HTTPRoute`s in the same namespace
into a single {{site.base_gateway}} Service. Enabling the feature gate can reduce the number of translated {{site.base_gateway}} Services.

The names of {{site.base_gateway}} Services will change if the feature gate is enabled.
You can refer to the [`HTTRoutes` reference page](/kubernetes-ingress-controller/faq/combining-httproutes/) for further details.
```
